### DISCUSSION TITLE
Dynamic text inside video.add_text does not update

### DISCUSSION THREAD

**Discussion by gavin2812** [savonet/liquidsoap#3072] [User Support]
Hi all,

I'm using liquidsoap 2.1.4 and I am trying to add dynamic text to a video using this:

```
v = video.add_text.ffmpeg.raw.filter(
      color=0xff0000,
      x=20,
      y=500,
      size=36,
      speed=0,
      graph=graph,
      screentext,
      v)
```
      
where `screentext` is this:

```
def update_time() =
	screentext = process.read("cat /home/test/mytext.txt")
	print(screentext)
end
thread.run(every=1.,update_time)
```
But even though the text updates correctly in the console, it does not update on the screen. I have it hardcoded at the beginning of the script as follows:

`screentext = process.read("cat /home/test/mytext.txt")
`

And on the screen it initially shows that correctly, but it does not update as per the timed function even though it does update in the console correctly.

Any ideas why this might be?

Thanks in advance.

**Comment by smimram** [comment]
You need a reference for this: the definition of `screentext` is otherwise local to the `update_time` function. It should thus be something like
```
screentext = ref("undefined")

v = video.add_text.ffmpeg.raw.filter(
      color=0xff0000,
      x=20,
      y=500,
      size=36,
      speed=0,
      graph=graph,
      {!screentext},
      v)

def update_time() =
	screentext := process.read("cat /home/test/mytext.txt")
	print(!screentext)
end
thread.run(every=1.,update_time)
```

**Reply by gavin2812** [reply]
Hi @smimram,
Thanks for responding.
Unfortunately that didn't work, it just outputs the word "undefined" on the screen and never changes, the text updates in console correctly, so I know the variable `!screentext` is updating (and therefore the `update_time` is working, it just as if it is not updating the variable `!screentext` inside the `video.add_text` section.
I've tried every variation of this I can think of but I just cannot get the text to update on screen.
Any further ideas? Or maybe an alternative way to do it?
Thanks in advance.

**Reply by smimram** [reply]
From [the doc](https://www.liquidsoap.info/doc-dev/reference.html#video.add_text.ffmpeg.raw), you actually have to call `on_change` method to update the text once it has changed with this operator. You should thus further change `update_time` to
```
def update_time() =
	screentext := process.read("cat /home/test/mytext.txt")
	print(!screentext)
        v.on_change()
end
```

**Reply by gavin2812** [reply]
@smimram thanks for your reply. I managed to get it working, I had to move the code around a bit, but ended up getting it working.

Thanks for your help, much appreciated!
